templates/navbar.njk

```diff
@@ -0,0 +1,77 @@
+#}
+
+{# 
+A macro that adds a menu navigation bar with the provided items.
+It receives:
+- signInHref: the URL of the sign in / register page
+- menuItems: a collection of objects representing individual menu items.
+    Each object must contain properties with the following keys to build an anchor:
+    - id: it's the anchor element id
+    - href: the URL of the page the menu item will point to
+    - displayText: the text to display in the anchor
+- userDisplayText: a text identifying the logged user, e.g. email address or name
+#}
+{% macro addNavbar(signInHref, menuItems, userDisplayText) %}
+    <div class="js-toggle-nav" id="global-nav">
+        <nav role="navigation" class="content" aria-label="{{ lang.mainNavbar or 'The main navigation bar' }}">
+            {% if userDisplayText %}
+                <a href="#navigation" class="js-header-toggle menu">{{ userDisplayText }}
+                    <span class="govuk-visually-hidden">{{ lang.clickToExpandOrCollapse or 'Click to expand or collapse' }}</span>
+                </a>
+            {% else %}
+                <a id="user-signin-mobile" href="{{ signInHref }}" class="sign-in govuk-link">{{ lang.signInOrRegister or 'Sign in / Register' }}</a>
+            {% endif %}
+
+            <ul id="navigation">
+                <li id="signed-in-user" class="user">
+                    {{ userDisplayText }}
+                </li>
+                {% for item in menuItems %}
+                    <li>
+                        <a class="govuk-link" id="{{ item.id }}" href="{{ item.href }}">{{ item.displayText }}</a>
+                    </li>
+                {% endfor %}
+            </ul>
+        </nav>
+    </div>
+{% endmacro %}
+
+{# 
+A macro that adds a menu navigation bar with predefined items.
+It receives:
+- userDisplayText: a text identifying the logged user, e.g. email address or name
+- chsMonitorGuiUrl: the URL for the 'Companies you follow' link
+#}
+{% macro addPredefinedNavbar(userDisplayText, chsMonitorGuiUrl) %}
+    {% set menuItems = [
+        { id: "your-companies", href: "/your-companies", displayText: lang.yourCompanies or "Your companies" },
+        { id: "your-filings", href: "/user/transactions", displayText: lang.yourFilings or "Your filings" },
+        { id: "companies-you-follow", href: chsMonitorGuiUrl, displayText: lang.companiesYouFollow or "Companies you follow" },
+        { id: "basket", href: "/basket", displayText: lang.basket or "Basket" },
+        { id: "your-details", href: "/user/account", displayText: lang.manageAccount or "Manage account" },
+        { id: "user-signout", href: "/signout", displayText: lang.signOut or "Sign out" }
+    ] %}
+
+    {% if displayAuthorisedAgent === "yes" %}
+        {% set manuItems = (menuItems.splice(1, 0, { id: "authorised-agent", href: "/authorised-agent", displayText: lang.authorisedAgent or "Authorised agent" }), menuItems) %}
+    {% endif %}
```

Comment:
I'm sure it works.
Had to go and check what the splice is doing there ..... thinking of the extendibility/Maintainability
could you check if it works something like adding any optional element straight into the array:
```
        ...
        { id: "your-companies", href: "/your-companies", displayText: lang.yourCompanies or "Your companies" },
        displayAuthorisedAgent == "yes" ? { id: "authorised-agent", href: "/authorised-agent", displayText: lang.authorisedAgent or "Authorised agent" } : null,
        { id: "your-filings", href: "/user/transactions", displayText: lang.yourFilings or "Your filings" },
        ...
```  
and then filter all the `null` out (so at line `69` something like:
```
    ] | filter(item => item) %}
```
).
Might you try and see if it works in your testing env and we can have a thought ?
If it works, that seems to me a little easier for people having to add any other optional element in the future.

Reply:
I refactored the code but had to find a suitable Nunjucks function since there is no filter function as you proposed and due to the poor documentation, I had to go to Nunjucks source code to find the correct argument to the reject() function. But I tested the solution and it works.